### PR TITLE
[Store] support segment level metrics(fix code format of #1029)

### PR DESCRIPTION
**Pull request by cocktail828**
# Description
This PR(which has already been integrated into the iflytek's Spark LLM) introduces segment-level memory metrics to track three key indicators for each segment:
- allocated: Actual memory allocated by the segment
- capacity: Maximum memory capacity of the segment
- ratio: Memory utilization ratio (allocated / capacity)

In production environments, monitoring memory usage at the segment granularity is critical for:
1. Gaining visibility into per-segment resource consumption (avoiding blind spots in global memory monitoring)
2. Identifying resource-heavy segments that may cause bottlenecks
3. Laying the foundation for subsequent load balancing optimizations

# Motivation & Background
Currently, the system only provides global memory metrics, which makes it difficult to pinpoint segment-specific memory issues in production. For large-scale deployments with multiple segments, understanding the memory distribution across individual segments is essential for stable operation.
This change addresses the gap by adding fine-grained metrics, enabling operators to:

- Proactively detect abnormal memory growth in specific segments
- Further more, optimize resource allocation based on real-world segment usage
- Troubleshoot memory-related issues more efficiently

# Future Roadmap
This segment-level memory metric is a foundational enhancement for follow-up work:
- Implement global load balancing based on segment memory utilization (e.g., migrating segments with high ratio to underutilized nodes)
- Build alerting mechanisms for abnormal memory usage (e.g., ratio exceeding threshold)

Implementation Details
- Added metrics tracking logic in the segment management module
- Added unit tests to verify metric accuracy

# Testing
All modifications are covered by tests, with all cases passing.

### CODE REVIEW

**Comment by cocktail828** [comment]
@ykwd Could you please review this PR? All code formatting issues have been resolved.

File: mooncake-store/tests/master_metrics_test.cpp [hunk 1]
```diff
@@ -28,12 +28,12 @@ TEST_F(MasterMetricsTest, InitialStatusTest) {
...
     ASSERT_EQ(metrics.get_allocated_mem_size(), 0);
     ASSERT_EQ(metrics.get_total_mem_capacity(), 0);
-    ASSERT_DOUBLE_EQ(metrics.get_global_mem_used_ratio(), 0.0);
+    ASSERT_DOUBLE_EQ(metrics.get_mem_used_ratio(), 0.0);
 
     // File Storage Metrics
     ASSERT_EQ(metrics.get_allocated_file_size(), 0);
     ASSERT_EQ(metrics.get_total_file_capacity(), 0);
-    ASSERT_DOUBLE_EQ(metrics.get_global_file_used_ratio(), 0.0);
+    ASSERT_DOUBLE_EQ(metrics.get_mem_used_ratio(), 0.0);
```
Comment (ykwd) [inline mooncake-store/tests/master_metrics_test.cpp]: By mistake? Seems like gemini ai also catched this. Please take a look.

**Comment by ykwd** [comment]
Thanks for the work. I have a question. The newly added interface parameters, such as get_mem_used_ratio(segment='xxx'), don’t seem to be called anywhere. Could you explain how this interface is intended to be invoked externally?

**Comment by cocktail828** [comment]
Thanks for your review and question!
The get_mem_used_ratio(segment='xxx') and other new segment-level metric interfaces are foundational. 
Background: In production, we’ve observed significant load imbalance during kvagent scaling—some nodes reach over 95% memory utilization (desperately needing GC) but our current GC strategy doesn’t account for this scenario.

These interfaces will serve two key purposes (per roadmap):
- Called by our existing metrics pipeline (e.g., Prometheus exporter) to feed data to Grafana for monitoring/alerting (handled by the metrics team; no changes needed here).
- Invoked by future GC logic to get real-time memory ratios, enabling targeted segment-level or even global GC for imbalanced nodes.

They’re not referenced yet as the consuming code (metrics integration/GC) isn’t part of this PR.

File: mooncake-store/include/allocator.h [hunk 2]
```diff
@@ -72,6 +74,8 @@ class AllocatedBuffer {
 
    private:
     std::weak_ptr<BufferAllocatorBase> allocator_;
+    // init segment name at startup
+    std::string segment_name_;
```
Comment (ykwd) [inline mooncake-store/include/allocator.h]: This class already has getSegmentName() api

**Comment by ykwd** [comment]
I see. That is a common requirement, and it’s already implemented in the current codebase.
There is a MasterService::QuerySegments method, which can be accessed via the HTTP endpoint /query_segment defined in rpc_service.cpp.

File: mooncake-store/include/master_metric_manager.h [hunk 3]
```diff
@@ -19,16 +19,16 @@ class MasterMetricManager {
...
-    void dec_total_mem_capacity(int64_t val = 1);
+    void inc_total_mem_capacity(const std::string& segment, int64_t val = 1);
+    void dec_total_mem_capacity(const std::string& segment, int64_t val = 1);
     void reset_total_mem_capacity();
-    int64_t get_allocated_mem_size();
-    int64_t get_total_mem_capacity();
-    double get_global_mem_used_ratio(void);
+    int64_t get_allocated_mem_size(const std::string& segment = "");
+    int64_t get_total_mem_capacity(const std::string& segment = "");
+    double get_mem_used_ratio(const std::string& segment = "");
```
Comment (ykwd) [inline mooncake-store/include/master_metric_manager.h]: For the three get_xxx functions (get_mem_used_ratio, get_total_mem_capacity, get_allocated_mem_size), it might be better to add new functions instead of modifying the existing ones, since retrieving memory size or other metrics for a specific segment differs in functionality and semantics from getting global metrics.

**Review by ykwd — COMMENTED** [review]
I just had an offline discussion with @xiaguan, which made me realize that making the change directly in the master metric manager allows it to be more easily accessed by Prometheus.
If that’s the case, then this PR makes sense.
I do have a few comments regarding the code.

File: mooncake-store/tests/master_metrics_test.cpp (hunk 1 above)
Comment (cocktail828) [reply mooncake-store/tests/master_metrics_test.cpp]: Got it – this must be an oversight. I haven’t modified any code related to the "file usage ratio".
I'll fix this right away.

File: mooncake-store/include/allocator.h (hunk 2 above)
Comment (cocktail828) [reply mooncake-store/include/allocator.h]: Great question! I actually tried using this API too, but during unit testing, I noticed that the ‘std::weak_ptr<BufferAllocatorBase> allocator_‘ might have already been released — for example, after an ‘unmount operation’. In such cases, we can no longer retrieve the segment's name.

File: mooncake-store/tests/master_metrics_test.cpp (hunk 1 above)
Comment (cocktail828) [reply mooncake-store/tests/master_metrics_test.cpp]: Fixed

File: mooncake-store/include/master_metric_manager.h (hunk 3 above)
Comment (cocktail828) [reply mooncake-store/include/master_metric_manager.h]: Fixed

**Comment by cocktail828** [comment]
@ykwd Thanks for your review.
Addressed test typos and implemented the requested segment-level metric get APIs per review feedback.

File: mooncake-store/src/segment.cpp [hunk 4]
```diff
@@ -199,12 +199,18 @@ ErrorCode ScopedSegmentAccess::CommitUnmountSegment(
...
+    if (segment != segment_manager_->mounted_segments_.end()) {
+        segment_name = segment->second.segment.name;
+    }
     // Remove from mounted_segments_
     segment_manager_->mounted_segments_.erase(segment_id);
 
     // Decrease the total capacity
     MasterMetricManager::instance().dec_total_mem_capacity(
-        metrics_dec_capacity);
+        segment_name, metrics_dec_capacity);
```
[the remark this answers is not in the document]
Comment (cocktail828) [reply mooncake-store/src/segment.cpp:213]: No need to handle the scenario where the segment is empty separately — the metrics internally include automatic fallback logic to ensure compatibility. We can ignore this part!

File: mooncake-store/include/allocator.h (hunk 2 above)
Comment (ykwd) [reply mooncake-store/include/allocator.h]: I see. Would it make sense to decrease the allocator’s current size entirely when the BufferAllocator is released?
When we optimize the master’s performance later, replica and allocatedBuffer will also be further simplified.

Comment (cocktail828) [reply mooncake-store/include/allocator.h]: This scenario is quite edge-case. Realistically, the segment has already been deallocated at this point, and its memory usage is no longer a concern. The reason for handling it this way is to maintain consistency in the test style and ensure the correctness of the test cases.
In my opinion, either approach works here—it’s irrelevant/makes no difference.

Comment (ykwd) [reply mooncake-store/include/allocator.h]: Agree. Both approaches work. The reason I recommend storing one fewer segment name is for performance considerations.

Comment (cocktail828) [reply mooncake-store/include/allocator.h]: OK, I'll make the adjustments this afternoon. I'll also add a comment to document the points we discussed for clarity.

Comment (ykwd) [reply mooncake-store/include/allocator.h]: Thanks

Comment (cocktail828) [reply mooncake-store/include/allocator.h]: Changes completed.
